pkg/api/rebalance.go

```diff
@@ -0,0 +1,147 @@
+	FastMode bool `param:"fast_mode,omitempty"`
+	// List of goals used to generate proposal, the default goals will be used if this parameter is not specified.
+	Goals []types.Goal `param:"goals,omitempty"`
+	// Whether to use Kafka assigner mode to generate proposals.
+	KafkaAssigner bool `param:"kafka_assigner,omitempty"`
+	// Replica movement strategies to use.
+	ReplicaMovementStrategies []types.ReplicaMovementStrategy `param:"replica_movement_strategies,omitempty"`
+	// Upper bound on the bandwidth in bytes per second used to move replicas.
+	ReplicationThrottle int64 `param:"replication_throttle,omitempty"`
+	// Review id for 2-step verification.
+	ReviewID int32 `param:"review_id,omitempty"`
+	// Whether to allow hard goals be skipped in proposal generation.
+	SkipHardGoalCheck bool `param:"skip_hard_goal_check,omitempty"`
+	// Whether to stop the ongoing execution (if any) and start executing the given request.
+	StopOngoingExecution bool `param:"stop_ongoing_execution,omitempty"`
+	// Whether to only use ready goals to generate proposal.
+	UseReadyDefaultGoals bool `param:"use_ready_default_goals,omitempty"`
+	// Return detailed state information.
+	Verbose bool `param:"verbose,omitempty"`
+	// Whether to calculate proposal from available valid partitions or valid windows.
+	DataFrom types.ProposalDataSource `param:"data_from,omitempty"`
+	// The upper bound of ongoing replica movements between disks within each broker.
+	ConcurrentIntraBrokerPartitionMovements int32 `param:"concurrent_intra_broker_partition_movements,omitempty"`
+	// Whether to ignore the cached proposal or not.
+	IgnoreProposalCache bool `param:"ignore_proposal_cache,omitempty"`
+	// Whether to balance load between disks within brokers (requires JBOD Kafka deployment).
+	RebalanceDisk bool `param:"rebalance_disk,omitempty"`
+}
+
+func (s RebalanceRequest) Validate() error {
+	if s.ConcurrentPartitionMovementsPerBroker == 0 {
+		return errors.New("number of concurrent partition movements per broker must be bigger then 0")
+	}
+
+	if s.ConcurrentLeaderMovements == 0 {
+		return errors.New("number of concurrent leader partition movements must be bigger then 0")
+	}
+
+	if s.ConcurrentIntraBrokerPartitionMovements == 0 {
+		return errors.New("number of concurrent intra broker partition movements must be bigger then 0")
+	}
+
+	return nil
+}
+
+func RebalanceRequestWithDefaults() *RebalanceRequest {
+	return &RebalanceRequest{
+		AllowCapacityEstimation:               true,
+		ConcurrentLeaderMovements:             defaultConcurrentLeaderMovements,
+		ConcurrentPartitionMovementsPerBroker: defaultConcurrentPartitionMovementsPerBroker,
+		ExecutionProgressCheckIntervalMs:      defaultExecutionProgressCheckIntervalMs,
+		DataFrom:                              types.ProposalDataSourceValidWindows,
+	}
+}
+
+type RebalanceResponse struct {
+	types.GenericResponse
+
+	Result *types.OptimizationResult
+}
+
+func (r *RebalanceResponse) UnmarshalResponse(resp *http.Response) error {
```

Comment:
I haven't thought this thoroughly, but I might have a suggestion regarding this function and how to eliminate the code duplication.

Instead of this function:
```golang
func (r *RebalanceResponse) UnmarshalResponse(resp *http.Response) error {
```
You can write something like:
```golang
func UnmarshalResponse(resp *http.Response, into *TargetResponse) error {
```
where
```
type TargetResponse interface {
  SetResult(types.OptimizationResult)
  SetProgress(types.ProgressResult)
  SetError(types.APIError)
}
```
can be implemented by the `*Response` types all around the `api` package. You still have to implement these (or move these implementations to `types.GenericsResponse` maybe?), but I feel like it's less code than the current `UnmarshalResponse`. 

I am not sure this works, just wanted to come up with a solution about the multiple `UnmarshalResponse` functions.  WDYT?

Reply:
Yeah. I was thinking about this as well. The type of the `.Result` attribute vary depending on the API Response. E.g. the `Rebalance` API returns a `types.OptimizationResult` while other APIs like `KafkaParitionLoad` API returns `types.PartitionLoadState`. So using a single interface would not help a lot with removing all the code duplication. I agree that there are ways to get a more optimal implementation here (groups APIs based on responses, etc), I also wanted to keep it clean by not adding lot off "magic" which could degrade the readability of the code.

I consider this as an enhancement and I think it should be a separate effort.

Reply:
Understood